### PR TITLE
WIP: Migrate wasmer-cli to clap from structopt

### PR DESCRIPTION
**Pull request by silwol**
# Description

# Review

- [ ] Add a short description of the change to the CHANGELOG.md file

### CODE REVIEW

File: rust-toolchain [hunk 1]
```diff
@@ -1 +0,0 @@
-1.59
```
Comment (syrusakbary) [inline rust-toolchain:1]: Undo this plz

**Review by syrusakbary — APPROVED** [review]
It looks good other than the removal of the rust-toolchain file.
Once it's fixed we should be good to merge

**Comment by silwol** [comment]
There's still a few TODOs in the change that I'd like to have clarified first, that's why it is still a draft MR. The removal of the toolchain file will not be part of the final version of this MR of course, as has nothing to do with it.

**Comment by epilys** [comment]
I'd like to see a binary size of clap and structopt as well, out of curiosity.

**Comment by ptitSeb** [comment]
What about `cli-compiler`?

**Comment by silwol** [comment]
@ptitSeb you're right, had a newer version of the branch on my disk, force-pushed it now.
@epilys will collect a few numbers regarding binary size.
Didn't expect the MR to get that much attention while still in draft state. :-)

**Comment by silwol** [comment]
@epilys here are the numbers for changes in binary size (measured on my x64 desktop computer with Debian 11, without llvm compiler enabled, but I think that shouldn't change a lot in this context).

For transparency, this is how I collected the information:

```
# make sure we start clean
rm -rf target

# build wasmer cli (cranelift + singlepass, no llvm)
cargo build --release --manifest-path lib/cli/Cargo.toml --features cranelift,singlepass --bin wasmer

# build wasmer-headless cli
cargo build --release --manifest-path lib/cli/Cargo.toml --features headless --bin wasmer-headless

# build cli-compiler
cargo build --release --manifest-path lib/cli-compiler/Cargo.toml

# show filesize before stripping
echo "before stripping"
ls -l target/release/wasmer{,-headless,-compiler}

# strip the binaries
strip target/release/wasmer{,-headless,-compiler}

# show filesize after stripping
echo "after stripping"
ls -l target/release/wasmer{,-headless,-compiler}
```

## Raw release binaries
All raw numbers in the tabe are sizes in bytes.
|                 | structopt | clap     | absolute change | relative change |
| --------------- | --------- | -------- | --------------- | --------------- |
| wasmer          |  18334632 | 18299344 |          -35288 |          -0.19% |
| wasmer-compiler |   4896048 |  4839632 |          -56416 |          -1.15% |
| wasmer-headless |  12309944 | 12268000 |          -41944 |          -0.34% |


## Stripped release binaries
All raw numbers in the tabe are sizes in bytes.
|                 | structopt | clap     | absolute change | relative change |
| --------------- | --------- | -------- | --------------- | --------------- |
| wasmer          |  11744824 | 11707960 |          -36864 |          -0.31% |
| wasmer-compiler |   1304984 |  1251736 |          -53248 |          -4.08% |
| wasmer-headless |   6780424 |  6739464 |          -40960 |          -0.60% |

**Comment by silwol** [comment]
Superseded by https://github.com/wasmerio/wasmer/pull/3079

**Pull request closed without merging**